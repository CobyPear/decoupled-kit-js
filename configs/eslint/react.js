/** @type {import('eslint').ESLint.ConfigData} */
module.exports = {
	plugins: ['react'],
	extends: [
		'./index.js',
		'plugin:react/recommended',
		'plugin:react-hooks/recommended',
		'plugin:react/jsx-runtime',
	],
	plugins: ['react-refresh'],
	settings: {
		react: {
			version: 'detect',
		},
	},
};
