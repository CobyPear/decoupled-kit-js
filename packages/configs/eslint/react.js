module.exports = {
	plugins: ['react'],
	extends: [
		'./index.js',
		'plugin:react/recommended',
		'plugin:react-hooks/recommended',
	],
	settings: {
		react: {
			version: 'detect',
		},
	},
};
