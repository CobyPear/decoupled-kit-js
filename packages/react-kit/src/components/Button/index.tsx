export { IconButton } from './IconButton';
export { Button } from './Button';
export * from './props';
