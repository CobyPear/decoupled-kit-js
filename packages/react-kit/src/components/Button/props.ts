import { Slot } from '@radix-ui/react-slot';

export type BaseButtonProps = Readonly<{
	/**
	 * Use the child element as the element rendered for the button.
	 * Useful for a styling Link components as buttons.
	 * @example
	 * ```tsx
	 * import Link from 'next/link';
	 * import { Button } from '@pantheon-systems/react-kit/components/Button'
	 *
	 * // ...
	 * <Button asChild>
	 * 	<Link href="/about">About</Link>
	 * </Button>
	 * ```
	 */
	asChild?: boolean;
	/**
	 * Button content
	 */
	children?: React.ReactNode;
	className?: string;
	/**
	 * other props like onClick etc.
	 */
	[key: string]: unknown;
}>;

export type ButtonProps = BaseButtonProps &
	Readonly<{
		/**
		 * Sometimes we want a button style but semantically need a link.
		 * @remarks Use asChild to use the child component as rendered element.
		 */
		Element?: 'button' | 'a';
		/**
		 * Button type
		 */
		type?: 'primary' | 'secondary';
		/**
		 * Button size
		 */
		size?: 'small' | 'large';
	}>;

export type BaseButtonElement = React.ElementRef<'button'> &
	React.ElementRef<typeof Slot>;

export type ButtonElement = BaseButtonElement & React.ElementRef<'a'>;
