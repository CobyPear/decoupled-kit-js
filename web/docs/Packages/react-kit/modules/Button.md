---
id: 'Button'
title: 'Module: Button'
sidebar_label: 'Button'
sidebar_position: 0
custom_edit_url: null
---

## Type Aliases

### BaseButtonElement

Ƭ **BaseButtonElement**: `React.ElementRef`<`"button"`\> &
`React.ElementRef`<typeof `Slot`\>

#### Defined in

[packages/react-kit/src/components/Button/props.ts:47](https://github.com/pantheon-systems/decoupled-kit-js/blob/9ff365d0/packages/react-kit/src/components/Button/props.ts#L47)

---

### BaseButtonProps

Ƭ **BaseButtonProps**: `Readonly`<{ `[key: string]`: `unknown`; `asChild?`:
`boolean` ; `children?`: `React.ReactNode` ; `className?`: `string` }\>

#### Defined in

[packages/react-kit/src/components/Button/props.ts:3](https://github.com/pantheon-systems/decoupled-kit-js/blob/9ff365d0/packages/react-kit/src/components/Button/props.ts#L3)

---

### ButtonElement

Ƭ **ButtonElement**: [`BaseButtonElement`](Button.md#basebuttonelement) &
`React.ElementRef`<`"a"`\>

#### Defined in

[packages/react-kit/src/components/Button/props.ts:50](https://github.com/pantheon-systems/decoupled-kit-js/blob/9ff365d0/packages/react-kit/src/components/Button/props.ts#L50)

---

### ButtonProps

Ƭ **ButtonProps**: [`BaseButtonProps`](Button.md#basebuttonprops) & `Readonly`<{
`Element?`: `"button"` \| `"a"` ; `size?`: `"small"` \| `"large"` ; `type?`:
`"primary"` \| `"secondary"` }\>

#### Defined in

[packages/react-kit/src/components/Button/props.ts:30](https://github.com/pantheon-systems/decoupled-kit-js/blob/9ff365d0/packages/react-kit/src/components/Button/props.ts#L30)

## Functions

### Button

▸ **Button**(`props`): `ReactNode`

#### Parameters

| Name    | Type                                                                                                                    |
| :------ | :---------------------------------------------------------------------------------------------------------------------- |
| `props` | `Omit`<[`ButtonProps`](Button.md#buttonprops), `"ref"`\> & `RefAttributes`<[`ButtonElement`](Button.md#buttonelement)\> |

#### Returns

`ReactNode`

**`See`**

[https://live-storybook-react-kit.appa.pantheon.site/?path=/docs/button--docs](https://live-storybook-react-kit.appa.pantheon.site/?path=/docs/button--docs)

#### Defined in

node_modules/.pnpm/@types+react@18.2.28/node_modules/@types/react/index.d.ts:427

---

### IconButton

▸ **IconButton**(`props`): `ReactNode`

#### Parameters

| Name    | Type                                                                                                                                                                                                            |
| :------ | :-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `props` | `Omit`<`Readonly`<{ `[key: string]`: `unknown`; `asChild?`: `boolean` ; `children?`: `ReactNode` ; `className?`: `string` }\>, `"ref"`\> & `RefAttributes`<[`BaseButtonElement`](Button.md#basebuttonelement)\> |

#### Returns

`ReactNode`

**`See`**

[https://live-storybook-react-kit.appa.pantheon.site/?path=/docs/button--docs](https://live-storybook-react-kit.appa.pantheon.site/?path=/docs/button--docs)

#### Defined in

node_modules/.pnpm/@types+react@18.2.28/node_modules/@types/react/index.d.ts:427
