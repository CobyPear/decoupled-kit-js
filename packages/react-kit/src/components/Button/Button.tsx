import { Slot } from '@radix-ui/react-slot';
import clsx from 'clsx';
import React from 'react';
import { ButtonElement, ButtonProps } from './props';

/**
 *
 * @see {@link https://live-storybook-react-kit.appa.pantheon.site/?path=/docs/button--docs}
 * @param ButtonProps - {@link ButtonProps}
 */
export const Button = React.forwardRef<ButtonElement, ButtonProps>(
	(
		{
			Element = 'button',
			type = 'primary',
			size = 'small',
			children,
			asChild,
			className,
			...props
		},
		forwardedRef,
	) => {
		const Component = asChild ? Slot : Element;

		const BUTTON_STYLES = {
			base: 'rk-btn rk-capitalize',
			type: {
				primary: clsx(
					'rk-bg-neutral-900 rk-text-white rk-rounded-lg hover:rk-bg-neutral-700 hover:rk-border-neutral-700 hover:rk-text-white active:rk-bg-neutral-500',
				),
				secondary: clsx(
					'rk-bg-white rk-text-neutral-900 rk-rounded-lg rk-btn-outline hover:rk-bg-neutral-100 hover:rk-text-neutral-900 hover:rk-border-neutral-900 active:rk-bg-neutral-300',
				),
			},
			size: {
				small: clsx('rk-px-4 rk-h-10'),
				large: clsx('rk-px-8 rk-h-12'),
			},
		};

		return (
			<Component
				ref={forwardedRef}
				className={clsx(
					BUTTON_STYLES.base,
					BUTTON_STYLES.type[type],
					BUTTON_STYLES.size[size],
					className,
				)}
				{...props}
			>
				{children}
			</Component>
		);
	},
);

Button.displayName = 'Button';
