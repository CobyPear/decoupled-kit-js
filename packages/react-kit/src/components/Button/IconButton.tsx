import { Slot } from '@radix-ui/react-slot';
import clsx from 'clsx';
import React from 'react';
import { BaseButtonElement, BaseButtonProps } from './props';

/**
 *
 * @see {@link https://live-storybook-react-kit.appa.pantheon.site/?path=/docs/button--docs}
 * @param IconButtonProps - {@link IconButtonProps}
 */
export const IconButton = React.forwardRef<BaseButtonElement, BaseButtonProps>(
	({ children, asChild, className, ...props }, forwardedRef) => {
		const Component = asChild ? Slot : 'button';
		return (
			<Component
				ref={forwardedRef}
				className={clsx(
					'rk-btn',
					'rk-h-10 rk-max-h-10 rk-min-h-fit rk-rounded-full rk-border-none rk-bg-white rk-p-2 rk-text-black	hover:rk-bg-neutral-100 hover:rk-text-neutral-900 active:rk-bg-neutral-300 [&>*:first-child]:rk-w-6',
					className,
				)}
				{...props}
			>
				{children}
			</Component>
		);
	},
);

IconButton.displayName = 'IconButton';
