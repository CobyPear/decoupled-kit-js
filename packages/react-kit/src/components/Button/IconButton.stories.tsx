import CloseIcon from '@assets/icons/close.svg';
import HamburgerMenu from '@assets/icons/hamburger-menu.svg';
import SearchIcon from '@assets/icons/search.svg';
import { IconButton } from '@components/Button';
import type { Meta, StoryObj } from '@storybook/react';

const Icon = ({
	src,
	alt,
	ariaLabel,
}: { [key in 'src' | 'alt' | 'ariaLabel']?: string }) => {
	return <img src={src} alt={alt} aria-label={ariaLabel} width="16" />;
};

const SearchIconComponent = () => (
	<Icon src={SearchIcon} key="2" alt="Search" ariaLabel="search" />
);
const CloseIconComponent = () => (
	<Icon src={CloseIcon} key="2" alt="Close" ariaLabel="close" />
);
const HMIconComponent = () => (
	<Icon src={HamburgerMenu} key="4" alt="Menu" ariaLabel="open menu" />
);

const meta: Meta<typeof IconButton> = {
	title: 'Button/IconButton',
	component: IconButton,
	parameters: {
		layout: 'centered',
	},
	tags: ['autodocs'],
	argTypes: {
		children: {
			control: {
				type: 'select',
			},
			options: ['Search', 'Close', 'Hamburger_Menu'],
			defaultValue: 'Search',
			mapping: {
				Text: 'Button',
				Search: <SearchIconComponent />,
				Close: <CloseIconComponent />,
				Hamburger_Menu: <HMIconComponent />,
			},
		},
		className: {
			if: { arg: 'type', neq: 'icon' },
			control: {
				type: 'text',
			},
		},
	},
};

export default meta;

type Story = StoryObj<typeof meta>;

export const Search: Story = {
	args: {
		children: <SearchIconComponent />,
	},
};
export const Close: Story = {
	args: {
		children: <CloseIconComponent />,
	},
};
export const Hamburger_Menu: Story = {
	args: {
		children: <HMIconComponent />,
	},
};
