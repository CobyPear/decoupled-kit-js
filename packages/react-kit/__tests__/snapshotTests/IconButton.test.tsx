import { IconButton } from '@components/Button';
import { render } from '@testing-library/react';

describe('<IconButton />', () => {
	it('should render an IconButton', () => {
		const { asFragment } = render(
			<IconButton>
				<img src={'/mock/search/icon.svg'} alt="Search" />
			</IconButton>,
		);

		expect(asFragment()).toMatchSnapshot();
	});
});
