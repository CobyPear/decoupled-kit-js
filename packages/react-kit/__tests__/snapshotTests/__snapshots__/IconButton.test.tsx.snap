// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<IconButton /> > should render an IconButton 1`] = `
<DocumentFragment>
  <button
    class="rk-btn rk-h-10 rk-max-h-10 rk-min-h-fit rk-rounded-full rk-border-none rk-bg-white rk-p-2 rk-text-black	hover:rk-bg-neutral-100 hover:rk-text-neutral-900 active:rk-bg-neutral-300 [&>*:first-child]:rk-w-6"
  >
    <img
      alt="Search"
      src="/mock/search/icon.svg"
    />
  </button>
</DocumentFragment>
`;
